src/server/pfs/s3/bucket.go

```diff
@@ -0,0 +1,359 @@
+	return bucketHandler{pc: pc}
+}
+
+func (h bucketHandler) location(w http.ResponseWriter, r *http.Request) {
+	repo, branch := bucketArgs(w, r)
+	
+	_, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/xml")
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte(locationSource))
+}
+
+func (h bucketHandler) get(w http.ResponseWriter, r *http.Request) {
+	repo, branch := bucketArgs(w, r)
+
+	// ensure the branch exists and has a head
+	branchInfo, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+
+	maxKeys := defaultMaxKeys
+	maxKeysStr := r.FormValue("max-keys")
+	if maxKeysStr != "" {
+		i, err := strconv.Atoi(maxKeysStr)
+		if err != nil || i < 0 || i > defaultMaxKeys {
+			invalidArgument(w, r)
+			return
+		}
+		maxKeys = i	
+	}
+
+	delimiter := r.FormValue("delimiter")
+	if delimiter != "" && delimiter != "/" {
+		invalidDelimiterError(w, r)
+		return
+	}
+
+	result := &ListBucketResult{
+		Name:        repo,
+		Prefix:      r.FormValue("prefix"),
+		Marker:      r.FormValue("marker"),
+		Delimiter:   delimiter,
+		MaxKeys:     maxKeys,
+		IsTruncated: false,
+	}
+
+	if branchInfo.Head == nil {
+		// if there's no head commit, just print an empty list of files
+		writeXML(w, r, http.StatusOK, result)
+	} else if delimiter == "" {
+		h.listRecursive(w, r, result, branch)
+	} else {
+		h.list(w, r, result, branch)
+	}
+}
+
+func (h bucketHandler) listRecursive(w http.ResponseWriter, r *http.Request, result *ListBucketResult, branch string) {
+	err := h.pc.Walk(result.Name, branch, filepath.Dir(result.Prefix), func(fileInfo *pfs.FileInfo) error {
+		origFilePath := fileInfo.File.Path
+		fileInfo = updateFileInfo(branch, result.Marker, result.Prefix, fileInfo)
+		if fileInfo == nil || fileInfo.FileType == pfs.FileType_DIR {
+			return nil
+		}
+		if result.isFull() {
+			if result.MaxKeys > 0 {
+				result.IsTruncated = true
+			}
+			return errutil.ErrBreak
+		}
+
+		meta, err := getMeta(h.pc, result.Name, branch, origFilePath)
+		if err != nil {
+			return err
+		}
+		etag := ""
+		if meta != nil {
+			etag = meta.MD5
+		}
+		
+		contents, err := newContents(fileInfo, etag)
+		if err != nil {
+			return err
+		}
+		result.Contents = append(result.Contents, contents)
+		
+		return nil
+	})
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+
+	setNextMarker(result)
+	writeXML(w, r, http.StatusOK, result)
+}
+
+func (h bucketHandler) list(w http.ResponseWriter, r *http.Request, result *ListBucketResult, branch string) {
+	pattern := fmt.Sprintf("%s*", glob.QuoteMeta(result.Prefix))
+	fileInfos, err := h.pc.GlobFile(result.Name, branch, pattern)
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+
+	for _, fileInfo := range fileInfos {
+		origFilePath := fileInfo.File.Path
+		fileInfo = updateFileInfo(branch, result.Marker, result.Prefix, fileInfo)
```

Comment:
`updateFileInfo` seems like a bit of an odd name for this, not sure what a better one would be, maybe `convertFileInfo`?. Also the fact that it updates the passed `FileInfo` in place and returns it seems a little bit weird to me. Reading just the signature of the function I'd expect it to be making a copy.

Reply:
Because I was able to nix `listRecursive`, `updateFileInfo` is no longer needed